File: 백준/Silver/9012. 괄호/괄호.py
def is_valid_parenthesis(string):
    stack = []
    for char in string:
        if char == '(':
            stack.append(char)
        elif char == ')':
            if not stack or stack.pop() != '(':
                return False
    return not stack

def main():
    t = int(input())
    for _ in range(t):
        string = input()
        print("YES" if is_valid_parenthesis(string) else "NO")

if __name__ == "__main__":
    main()

File: 백준/Silver/9012. 괄호/괄호.java
//괄호 9012

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Stack;

class Main {

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();

    int tc = Integer.parseInt(br.readLine());
    for (int i = 0; i < tc; i++) {
      String line = br.readLine();
      if (isBalanced(line)) {
        sb.append("YES\n");
      } else {
        sb.append("NO\n");
      }
    }

    System.out.print(sb.toString());
    br.close();
  }

  private static boolean isBalanced(String line) {
    Stack<Character> bracketStack = new Stack<>();
    HashMap<Character, Character> bracketDict = new HashMap<>();
    bracketDict.put('(', ')');

    for (char c : line.toCharArray()) {
      if (c == '(') {
        bracketStack.push(c);
      } else if (c == ')') {
        if (bracketStack.isEmpty()) {
          return false;
        }
        char openBracket = bracketStack.pop();
        if (bracketDict.get(openBracket) != c) {
          return false;
        }
      }
    }
    return bracketStack.isEmpty();
  }
}
